### PR TITLE
Adding /assets API to v3 package

### PR DESCRIPTION
**Pull request by ngaglani-tenable**
- This PR contains the proposed base structure for code for v3 resources
- Current structure is something like
```
io/
   __init__.py
	v3/
            __init__.py
            base/
                endpoints/
	                uw.py
                iterators/
                schema/
	                analysis.py
	                exports.py
            assets.py
```

- Previous usage
```tio.assets.list()```
- Updated usage
```tio.v3.assets.list()```

### CODE REVIEW

File: tenable/io/__init__.py [hunk 1]
```diff
@@ -420,3 +421,11 @@ def workbenches(self):
         :doc:`Tenable.io Workbenches APIs <workbenches>`.
         '''
         return WorkbenchesAPI(self)
+
+    @property
+    def v3(self):
+        '''
+        The interface object for the v3 API's
+        :doc:`Tenable.io v3 APIs`.
```
Comment (SteveMcGrath) [inline tenable/io/__init__.py:429]: docstring doesn't link to the appropriate doc page.

File: tenable/io/v3/__init__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    """
+    This will contain property for all resources/app under io
+    i.e Container Security, Web Application Security.
+    """
+
+    # pylint: disable=too-few-public-methods
+    @property
+    def assets(self):
+        """
+        The interface object for the Assets API's
```
Comment (SteveMcGrath) [inline tenable/io/v3/__init__.py:20]: docstring doesn't link to the appropriate doc page.

File: tenable/io/v3/__init__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+"""
+Version3API
+===========
```
Comment (SteveMcGrath) [inline tenable/io/v3/__init__.py:3]: docstring doesn't include the relevant toctree and define what the v3 API are.

File: tenable/io/v3/assets.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+            :obj:`list`:
+                List of asset records.
+
+        Examples:
+            >>> for asset in tio.v3.assets.list():
+            ...     pprint(asset)
+        """
+        return self._api.get("assets").json()["assets"]
+
+    def delete(self, uuid):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:38]: no type hints

File: tenable/io/v3/assets.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+
+        Examples:
+            >>> asset_id = '00000000-0000-0000-0000-000000000000'
+            >>> tio.v3.workbenches.asset_delete(asset_id)
+        """
+        raise NotImplementedError(
+            "Delete has not been implemented as it depends on workbenches"
+        )
+
+    def details(self, obj_id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:58]: no type hints

File: tenable/io/v3/assets.py [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+        Returns:
+            :obj:`dict`:
+                Asset resource definition.
+
+        Examples:
+            >>> asset = tio.v3.assets.details(
+            ...     '00000000-0000-0000-0000-000000000000')
+        """
+        check_obj_id = self._check("obj_id", obj_id, str)
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:76]: _check usage should be deprecated

File: tenable/io/v3/assets.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+.. autoclass:: AssetsAPI
+    :members:
+"""
+from tenable.io.v3.base.endpoints.uw import UWBaseEndpoint
+
+
+class AssetsAPI(UWBaseEndpoint):
+    """
+    This will contain all methods related to Assets
+    """
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:20]: An attribute `_path` should be defined on the class with the string value of "v3/assets".  This refers to the base path that this resource is on.

Comment (SteveMcGrath) [reply tenable/io/v3/assets.py:20]: An attribute `_conv_json` should be defined with a bool value of `True`.  This informs the HTTP method methods to attempt automatic conversion of json responses.

File: tenable/io/v3/assets.py [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+
+        Returns:
+            :obj:`list`:
+                List of asset records.
+
+        Examples:
+            >>> for asset in tio.v3.assets.list():
+            ...     pprint(asset)
+        """
+        return self._api.get("assets").json()["assets"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:36]: Should use `self._get()['assets']` instead.

File: tenable/io/v3/assets.py [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+        Returns:
+            :obj:`dict`:
+                Asset resource definition.
+
+        Examples:
+            >>> asset = tio.v3.assets.details(
+            ...     '00000000-0000-0000-0000-000000000000')
+        """
+        check_obj_id = self._check("obj_id", obj_id, str)
+        return self._api.get(f"assets/{check_obj_id}").json()
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:77]: should use `self._get(obj_id)` instead.

File: tenable/io/v3/assets.py [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+
+    def assign_tags(self, action, assets, tags):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:79]: no type hints

File: tenable/io/v3/assets.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+            :obj:`dict`:
+                Asset resource definition.
+
+        Examples:
+            >>> asset = tio.assets.tags(
+            ...     '00000000-0000-0000-0000-000000000000')
+        """
+        raise NotImplementedError("Not implemented yet as it depends on tags")
+
+    def asset_import(self, source, *assets):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:124]: no type hints

File: tenable/io/v3/assets.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        # We will likely want to perform some more stringent checking of the
+        # asset resources that are being defined, however a simple type check
+        # should suffice for now.
+        return self._api.post(
+            "import/assets",
+            json={
+                "assets": [self._check("asset", i, dict) for i in assets],
+                "source": self._check("source", source, str),
+            },
+        ).json()["asset_import_job_uuid"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:179]: will the API endpoint path change here in v3?  Also we shouldn't use check.

File: tenable/io/v3/assets.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+            :obj:`list`:
+                List of job records.
+
+        Examples:
+            >>> for job in tio.assets.list_import_jobs():
+            ...     pprint(job)
+        """
+        return self._api.get("import/asset-jobs").json()["asset_import_jobs"]
+
+    def import_job_details(self, uuid):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:197]: type hinting

File: tenable/io/v3/assets.py [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+            :obj:`dict`:
+                The job Resource record.
+
+        Examples:
+            >>> job = tio.assets.import_job_details(
+            ...     '00000000-0000-0000-0000-000000000000')
+            >>> pprint(job)
+        """
+        check_uuid = self._check("uuid", uuid, str)
+        return self._api.get(f"import/asset-jobs/{check_uuid}").json()
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:216]: shouldn't use check.

File: tenable/io/v3/assets.py [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+
+    def move_assets(self, source, destination, targets):
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:218]: type hinting

File: tenable/io/v3/assets.py [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+        Examples:
+            >>> asset = tio.assets.move_assets('00000000-0000-0000-0000-000000000000',
+            ...         '10000000-0000-0000-0000-000000000001', ["127.0.0.1"])
+            >>> pprint(asset)
+        """
+        payload = {
+            "source": self._check("source", source, "uuid"),
+            "destination": self._check("destination", destination, "uuid"),
+            "targets": ",".join(self._check("targets", targets, list)),
+        }
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:244]: check usage should be deprecated

File: tenable/io/v3/assets.py [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+            ...     })
+        """
+        # We will likely want to perform some more stringent checking of the
+        # asset resources that are being defined, however a simple type check
+        # should suffice for now.
+        return self._api.post(
+            "import/assets",
+            json={
+                "assets": [self._check("asset", i, dict) for i in assets],
+                "source": self._check("source", source, str),
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets.py:177]: _check usage should be deprecated

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]
Generally it appears that the assets.py was just lifted and placed here instead of being refactored to match updated code structures.  Refer to the exports PR for inspiration.

https://github.com/tenable/pyTenable/pull/463

**Comment by ngaglani-tenable** [comment]
The changes have been updated in PR #12 . Hence, closing this PR.

**Pull request closed without merging**